compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -911,18 +911,21 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         item_segment: &hir::PathSegment<'_>,
     ) -> Ty<'tcx> {
         let substs = self.ast_path_substs_for_ty(span, did, item_segment);
-        let ty = self.tcx().at(span).type_of(did);
+
+        // Lazily compute this to avoid cycles under `lazy_type_alias`
+        let mut ty = None;
+        let mut ty = || *ty.get_or_insert_with(|| self.tcx().at(span).type_of(did));
 
         if matches!(self.tcx().def_kind(did), DefKind::TyAlias)
-            && (ty.skip_binder().has_opaque_types() || self.tcx().features().lazy_type_alias)
+            && (self.tcx().features().lazy_type_alias || ty().skip_binder().has_opaque_types())
```

Comment:
This lazy caching of the `ty` variable seems a bit convoluted imo. I think the control flow is clearer if we just duplicate the weak alias branch by first doing:

```
if tcx.features.lazy_type_alias {
  make the projection
} else {
  // comment going here explaining the query cycle condition being avoided with `lazy_type_alias`
  let ty = tcx.type_of(...);
  if ty.has_opaques() {
    // old comment mentioning opaques goes here
    make projection
  } else {
    ty.subst(..)
  }
}
```